synedrion/src/tools/secret.rs

```diff
@@ -39,22 +73,349 @@ impl<T: Zeroize> Debug for Secret<T> {
+
+    fn mul(mut self, other: &'a Secret<T>) -> Self::Output {
+        self *= other.expose_secret();
+        self
+    }
+}
+
+// Division
+
+impl<'a, T: Zeroize + DivAssign<&'a NonZero<T>>> DivAssign<&'a NonZero<T>> for Secret<T> {
+    fn div_assign(&mut self, other: &'a NonZero<T>) {
+        self.expose_secret_mut().div_assign(other)
+    }
+}
+
+impl<T: Zeroize + for<'a> DivAssign<&'a NonZero<T>>> Div<NonZero<T>> for Secret<T> {
+    type Output = Secret<T>;
+
+    fn div(mut self, other: NonZero<T>) -> Self::Output {
+        self /= &other;
+        self
+    }
+}
+
+// Remainder
+
+impl<'a, T: Zeroize + Clone + RemAssign<&'a NonZero<T>>> RemAssign<&'a NonZero<T>> for Secret<T> {
+    fn rem_assign(&mut self, other: &'a NonZero<T>) {
+        self.expose_secret_mut().rem_assign(other)
+    }
+}
+
+impl<'a, T: Zeroize + Clone + RemAssign<&'a NonZero<T>>> Rem<&'a NonZero<T>> for &Secret<T> {
+    type Output = Secret<T>;
+
+    fn rem(self, other: &'a NonZero<T>) -> Self::Output {
+        let mut result = self.clone();
+        result %= other;
+        result
+    }
+}
+
+// Summation
+
+impl<T: Zeroize + Clone + for<'a> AddAssign<&'a T> + Default> core::iter::Sum for Secret<T> {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.reduce(Add::add).unwrap_or(Secret::init_with(|| T::default()))
+    }
+}
+
+impl<'b, T: Zeroize + Clone + for<'a> AddAssign<&'a T> + Default> core::iter::Sum<&'b Secret<T>> for Secret<T> {
+    fn sum<I: Iterator<Item = &'b Secret<T>>>(iter: I) -> Self {
+        iter.fold(Secret::init_with(|| T::default()), |accum, x| accum + x)
+    }
+}
+
+impl Secret<Scalar> {
+    pub fn mul_by_generator(&self) -> Point {
+        self.expose_secret().mul_by_generator()
+    }
+
+    pub fn invert(&self) -> Option<Secret<Scalar>> {
+        Secret::maybe_init_with(|| Option::from(self.expose_secret().invert()))
+    }
+}
+
+impl Mul<Secret<Scalar>> for Point {
+    type Output = Point;
+    fn mul(self, scalar: Secret<Scalar>) -> Self::Output {
+        self * scalar.expose_secret()
+    }
+}
+
+impl Mul<&Secret<Scalar>> for Point {
+    type Output = Point;
+    fn mul(self, scalar: &Secret<Scalar>) -> Self::Output {
+        self * scalar.expose_secret()
+    }
+}
+
+impl Mul<Secret<Scalar>> for &Point {
+    type Output = Point;
+    fn mul(self, scalar: Secret<Scalar>) -> Self::Output {
+        self * scalar.expose_secret()
+    }
+}
+
+impl<T: Zeroize + Retrieve<Output: Zeroize + Clone>> Retrieve for Secret<T> {
+    type Output = Secret<T::Output>;
+    fn retrieve(&self) -> Self::Output {
+        Secret::init_with(|| self.expose_secret().retrieve())
+    }
+}
+
+impl<T: Zeroize> Secret<T> {
+    pub fn pow_bounded<V>(&self, exponent: &Bounded<V>) -> Self
+    where
+        T: Exponentiable<V>,
+        V: Integer + crypto_bigint::Bounded + Encoding + ConditionallySelectable,
+    {
+        // TODO: do we need to implement our own windowed exponentiation to hide the secret?
```

Comment:
What is it about `pow_[bounded, signed]` that worries you? Isn't the same concern valid for `mul`, `invert` and many other methods?

Reply:
The exponent will be put in a stack array when it's decomposed with a small radix for windowed exponentiation. So if it's secret, it's going to leave traces on the stack. With the multiplication there's less danger since `Uint`s implement `*Assign` traits which we use, so theoretically anything secret will be overwritten.

Reply:
This is a somewhat subtle point, would be good to have your comment here in the code too for our future selves. Handling hazmat is *hard*!

Reply:
Added the comment